server/src/main/java/org/opensearch/action/bulk/TransportBulkAction.java

```diff
@@ -644,25 +653,31 @@ public void onFailure(Exception e) {
                         // create failures for all relevant requests
                         for (BulkItemRequest request : requests) {
                             final String indexName = concreteIndices.getConcreteIndex(request.index()).getName();
-                            DocWriteRequest<?> docWriteRequest = request.request();
-                            responses.set(
+                            final DocWriteRequest<?> docWriteRequest = request.request();
+                            final BulkItemResponse bulkItemResponse = new BulkItemResponse(
                                 request.id(),
-                                new BulkItemResponse(
-                                    request.id(),
-                                    docWriteRequest.opType(),
-                                    new BulkItemResponse.Failure(indexName, docWriteRequest.id(), e)
-                                )
+                                docWriteRequest.opType(),
+                                new BulkItemResponse.Failure(indexName, docWriteRequest.id(), e)
                             );
+
+                            responses.set(request.id(), bulkItemResponse);
                         }
+
                         if (counter.decrementAndGet() == 0) {
                             finishHim();
                         }
                     }
 
                     private void finishHim() {
-                        listener.onResponse(
-                            new BulkResponse(responses.toArray(new BulkItemResponse[responses.length()]), buildTookInMillis(startTimeNanos))
+                        BulkItemResponse[] response = responses.toArray(new BulkItemResponse[responses.length()]);
+                        long buildTime = buildTookInMillis(startTimeNanos);
+
+                        indicesService.addDocStatusCounter(
+                            Arrays.stream(response)
+                                .collect(Collectors.groupingBy(x -> x.status().getStatusFamilyCode(), Collectors.counting()))
                         );
```

Comment:
Do we need this additional iteration? Can we instead count this within `onResponse, onFailure` and only update here with the values?